### PR TITLE
pkg/trace/api: experiment with trace decoding

### PR DESCRIPTION
**Pull request by knusbaum**
### What does this PR do?

This PR contains changes that experiment with trace decoding performance improvements.

### CODE REVIEW

File: pkg/trace/api/api.go [hunk 1]
```diff
@@ -562,10 +561,49 @@ func (r *HTTPReceiver) Languages() string {
 	return strings.Join(str, "|")
 }
 
-func decodeRequest(req *http.Request, dest msgp.Decodable) error {
+//var bufioReaderPool sync.Pool
```
Comment (knusbaum) [inline pkg/trace/api/api.go:564]: The bufioReaderPool was used for 1MB and 5MB read buffers

File: pkg/trace/api/api.go [hunk 2]
```diff
@@ -562,10 +561,49 @@ func (r *HTTPReceiver) Languages() string {
...
+//}
+//
+//func putBufioReader(br *bufio.Reader) {
+//	br.Reset(nil)
+//	bufioReaderPool.Put(br)
+//}
+
+func decodeRequest(req *http.Request, dest *pb.Traces) error {
+	//br := newBufioReader(req.Body)
+	//defer putBufioReader(br)
```
Comment (knusbaum) [inline pkg/trace/api/api.go:586]: for 1MB and 5MB tests, `br` was passed to `msgp.Decode`, as seen below.

File: pkg/trace/api/api.go [hunk 3]
```diff
@@ -562,10 +561,49 @@ func (r *HTTPReceiver) Languages() string {
...
 	switch mediaType := getMediaType(req); mediaType {
 	case "application/msgpack":
-		return msgp.Decode(req.Body, dest)
+		if req.ContentLength > 0 {
+			bs := make([]byte, req.ContentLength)
+			_, err := io.ReadFull(req.Body, bs)
+			if err != nil {
+				return err
+			}
+			return directDecodeTraces(bs, dest)
```
Comment (knusbaum) [inline pkg/trace/api/api.go:595]: This is the most optimal path, since we allocate the full buffer in one go.

File: pkg/trace/api/api.go [hunk 4]
```diff
@@ -562,10 +561,49 @@ func (r *HTTPReceiver) Languages() string {
...
+		}
+		panic("failed to decode request.")
+		// Normally we don't use ioutil.ReadAll, but req.Body is limited to maxRequestBodyLength
+		// with a LimitedReader.
+		bs, err := ioutil.ReadAll(req.Body)
```
Comment (knusbaum) [inline pkg/trace/api/api.go:600]: This is less optimal, but can be greatly improved from `ioutil.ReadAll`. `ioutil.ReadAll` uses a `bytes.Buffer`, which does not grow itself optimally for this use case. That can be optimized fairly easily.

**Comment by gbbr** [comment]
I managed to get this over your PR:

```
benchcmp pr.bench trim.bench 
benchmark                    old ns/op     new ns/op     delta
BenchmarkDecodeRequest-8     66728632      60679977      -9.06%

benchmark                    old allocs     new allocs     delta
BenchmarkDecodeRequest-8     501007         501007         +0.00%

benchmark                    old bytes     new bytes     delta
BenchmarkDecodeRequest-8     102551545     93655019      -8.68%
```

By simplifying the codec a bit (this requires tracer changes):

```diff
diff --git a/pkg/trace/api/api_direct_decode.go b/pkg/trace/api/api_direct_decode.go
index f8ab7bf8c..1c4c6d62e 100644
--- a/pkg/trace/api/api_direct_decode.go
+++ b/pkg/trace/api/api_direct_decode.go
@@ -1,45 +1,47 @@
 func directDecodeSpan(p []byte, z *pb.Span) (bs []byte, err error) {
-	var sz uint32
-	sz, bs, err = readMapHeader(p)
-	if err != nil {
-		return nil, err
-	}
-	for sz > 0 {
-		sz--
-		p, bs, err = readMapKeyPtr(bs)
-		if err != nil {
-			return nil, err
-		}
-		
-		switch msgp.UnsafeString(p) {
-		case "service":
+	bs = p
+	for i := uint32(0); i <= 11; i++ {
+		switch i {
+		case 0:
 			if bs[0] == mnil {
 				z.Service, err = "", nil
 				break
@@ -977,7 +969,7 @@ func directDecodeSpan(p []byte, z *pb.Span) (bs []byte, err error) {
 			if err != nil {
 				return
 			}
-		case "name":
+		case 1:
 			if bs[0] == mnil {
 				z.Name, err = "", nil
 				break
@@ -986,7 +978,7 @@ func directDecodeSpan(p []byte, z *pb.Span) (bs []byte, err error) {
 			if err != nil {
 				return
 			}
-		case "resource":
+		case 2:
 			if bs[0] == mnil {
 				z.Resource, err = "", nil
 				break
@@ -995,7 +987,7 @@ func directDecodeSpan(p []byte, z *pb.Span) (bs []byte, err error) {
 			if err != nil {
 				return
 			}
-		case "trace_id":
+		case 3:
 			if bs[0] == mnil {
 				z.TraceID, err = 0, nil
 				break
@@ -1004,7 +996,7 @@ func directDecodeSpan(p []byte, z *pb.Span) (bs []byte, err error) {
 			if err != nil {
 				return
 			}
-		case "span_id":
+		case 4:
 			if bs[0] == mnil {
 				z.SpanID, err = 0, nil
 				break
@@ -1013,7 +1005,7 @@ func directDecodeSpan(p []byte, z *pb.Span) (bs []byte, err error) {
 			if err != nil {
 				return
 			}
-		case "start":
+		case 5:
 			if bs[0] == mnil {
 				z.Start, err = 0, nil
 				break
@@ -1022,7 +1014,7 @@ func directDecodeSpan(p []byte, z *pb.Span) (bs []byte, err error) {
 			if err != nil {
 				return
 			}
-		case "duration":
+		case 6:
 			if bs[0] == mnil {
 				z.Duration, err = 0, nil
 				break
@@ -1031,7 +1023,7 @@ func directDecodeSpan(p []byte, z *pb.Span) (bs []byte, err error) {
 			if err != nil {
 				return
 			}
-		case "error":
+		case 7:
 			if bs[0] == mnil {
 				z.Error, err = 0, nil
 				break
@@ -1040,7 +1032,7 @@ func directDecodeSpan(p []byte, z *pb.Span) (bs []byte, err error) {
 			if err != nil {
 				return
 			}
-		case "meta":
+		case 8:
 			if bs[0] == mnil {
 				z.Meta, err = nil, nil
 				break
@@ -1072,7 +1064,7 @@ func directDecodeSpan(p []byte, z *pb.Span) (bs []byte, err error) {
 				}
 				z.Meta[zxvk] = zbzg
 			}
-		case "metrics":
+		case 9:
 			if bs[0] == mnil {
 				z.Metrics, err = nil, nil
 				break
@@ -1104,7 +1096,7 @@ func directDecodeSpan(p []byte, z *pb.Span) (bs []byte, err error) {
 				}
 				z.Metrics[zbai] = zcmr
 			}
-		case "parent_id":
+		case 10:
 			if bs[0] == mnil {
 				z.ParentID, err = 0, nil
 				break
@@ -1114,7 +1106,7 @@ func directDecodeSpan(p []byte, z *pb.Span) (bs []byte, err error) {
 			if err != nil {
 				return
 			}
-		case "type":
+		case 11:
 			if bs[0] == mnil {
 				z.Type, err = "", nil
 				break
@@ -1125,10 +1117,7 @@ func directDecodeSpan(p []byte, z *pb.Span) (bs []byte, err error) {
 				return
 			}
 		default:
-			bs, err = skip(bs)
-			if err != nil {
-				return
-			}
+			return bs, errors.New("unexpected entries in span map")
 		}
 	}
 	return
@@ -1148,7 +1137,7 @@ func directDecodeSpan(p []byte, z *pb.Span) (bs []byte, err error) {
 //		return err
 //	}
 
-func directDecodeTraces(bs []byte, ts *pb.Traces) error {	
+func directDecodeTraces(bs []byte, ts *pb.Traces) error {
 	//var sz uint32
 	//pb.Traces starts with an array header.
 	//fmt.Println("DIRECTDECODETRACES")
@@ -1189,4 +1178,4 @@ func directDecodeTraces(bs []byte, ts *pb.Traces) error {
 		}
 	}
 	return nil
-}
\ No newline at end of file
+}
diff --git a/pkg/trace/api/api_test.go b/pkg/trace/api/api_test.go
index 78d149ba3..3f8282665 100644
--- a/pkg/trace/api/api_test.go
+++ b/pkg/trace/api/api_test.go
@@ -557,6 +557,7 @@ func BenchmarkHandleTracesFromMultipleApps(b *testing.B) {
 	// benchmark
 	b.ResetTimer()
 	b.ReportAllocs()
+	b.SetBytes(int64(buf.Len()))
 	for n := 0; n < b.N; n++ {
 		b.StopTimer()
 		// consume the traces channel without doing anything
@@ -603,6 +604,17 @@ func BenchmarkDecoderJSON(b *testing.B) {
 	}
 }
 
+func TestNewEncode(t *testing.T) {
+	assert := assert.New(t)
+	want := testutil.GetTestTraces(1000, 100, true)
+	var buf bytes.Buffer
+	err := msgp.Encode(&buf, want)
+	assert.NoError(err)
+	var traces pb.Traces
+	directDecodeTraces(buf.Bytes(), &traces)
+	assert.Equal(want, traces)
+}
+
 func BenchmarkDecodeRequest(b *testing.B) {
 	assert := assert.New(b)
 	f, err := ioutil.TempFile("", "decodeRequest-*") //(f *os.File, err error)
@@ -630,7 +642,7 @@ func BenchmarkDecodeRequest(b *testing.B) {
 
 		b.StartTimer()
 		err = decodeRequest(req, &traces)
-		assert.NoError(err)		
+		assert.NoError(err)
 	}
 }
 
diff --git a/pkg/trace/pb/span_gen.go b/pkg/trace/pb/span_gen.go
index e8a9ed1c8..8a948b6d0 100644
--- a/pkg/trace/pb/span_gen.go
+++ b/pkg/trace/pb/span_gen.go
@@ -206,82 +206,82 @@ func (z *Span) DecodeMsg(dc *msgp.Reader) (err error) {
 func (z *Span) EncodeMsg(en *msgp.Writer) (err error) {
 	// map header, size 12
 	// write "service"
-	err = en.Append(0x8c, 0xa7, 0x73, 0x65, 0x72, 0x76, 0x69, 0x63, 0x65)
-	if err != nil {
-		return err
-	}
+	//err = en.Append(0x8c, 0xa7, 0x73, 0x65, 0x72, 0x76, 0x69, 0x63, 0x65)
+	//if err != nil {
+	//return err
+	//}
 	err = en.WriteString(z.Service)
 	if err != nil {
 		return
 	}
 	// write "name"
-	err = en.Append(0xa4, 0x6e, 0x61, 0x6d, 0x65)
-	if err != nil {
-		return err
-	}
+	//err = en.Append(0xa4, 0x6e, 0x61, 0x6d, 0x65)
+	//if err != nil {
+	//return err
+	//}
 	err = en.WriteString(z.Name)
 	if err != nil {
 		return
 	}
 	// write "resource"
-	err = en.Append(0xa8, 0x72, 0x65, 0x73, 0x6f, 0x75, 0x72, 0x63, 0x65)
-	if err != nil {
-		return err
-	}
+	//err = en.Append(0xa8, 0x72, 0x65, 0x73, 0x6f, 0x75, 0x72, 0x63, 0x65)
+	//if err != nil {
+	//return err
+	//}
 	err = en.WriteString(z.Resource)
 	if err != nil {
 		return
 	}
 	// write "trace_id"
-	err = en.Append(0xa8, 0x74, 0x72, 0x61, 0x63, 0x65, 0x5f, 0x69, 0x64)
-	if err != nil {
-		return err
-	}
+	//err = en.Append(0xa8, 0x74, 0x72, 0x61, 0x63, 0x65, 0x5f, 0x69, 0x64)
+	//if err != nil {
+	//return err
+	//}
 	err = en.WriteUint64(z.TraceID)
 	if err != nil {
 		return
 	}
 	// write "span_id"
-	err = en.Append(0xa7, 0x73, 0x70, 0x61, 0x6e, 0x5f, 0x69, 0x64)
-	if err != nil {
-		return err
-	}
+	//err = en.Append(0xa7, 0x73, 0x70, 0x61, 0x6e, 0x5f, 0x69, 0x64)
+	//if err != nil {
+	//return err
+	//}
 	err = en.WriteUint64(z.SpanID)
 	if err != nil {
 		return
 	}
 	// write "start"
-	err = en.Append(0xa5, 0x73, 0x74, 0x61, 0x72, 0x74)
-	if err != nil {
-		return err
-	}
+	//err = en.Append(0xa5, 0x73, 0x74, 0x61, 0x72, 0x74)
+	//if err != nil {
+	//return err
+	//}
 	err = en.WriteInt64(z.Start)
 	if err != nil {
 		return
 	}
 	// write "duration"
-	err = en.Append(0xa8, 0x64, 0x75, 0x72, 0x61, 0x74, 0x69, 0x6f, 0x6e)
-	if err != nil {
-		return err
-	}
+	//err = en.Append(0xa8, 0x64, 0x75, 0x72, 0x61, 0x74, 0x69, 0x6f, 0x6e)
+	//if err != nil {
+	//return err
+	//}
 	err = en.WriteInt64(z.Duration)
 	if err != nil {
 		return
 	}
 	// write "error"
-	err = en.Append(0xa5, 0x65, 0x72, 0x72, 0x6f, 0x72)
-	if err != nil {
-		return err
-	}
+	//err = en.Append(0xa5, 0x65, 0x72, 0x72, 0x6f, 0x72)
+	//if err != nil {
+	//return err
+	//}
 	err = en.WriteInt32(z.Error)
 	if err != nil {
 		return
 	}
 	// write "meta"
-	err = en.Append(0xa4, 0x6d, 0x65, 0x74, 0x61)
-	if err != nil {
-		return err
-	}
+	//err = en.Append(0xa4, 0x6d, 0x65, 0x74, 0x61)
+	//if err != nil {
+	//return err
+	//}
 	err = en.WriteMapHeader(uint32(len(z.Meta)))
 	if err != nil {
 		return
@@ -297,10 +297,10 @@ func (z *Span) EncodeMsg(en *msgp.Writer) (err error) {
 		}
 	}
 	// write "metrics"
-	err = en.Append(0xa7, 0x6d, 0x65, 0x74, 0x72, 0x69, 0x63, 0x73)
-	if err != nil {
-		return err
-	}
+	//err = en.Append(0xa7, 0x6d, 0x65, 0x74, 0x72, 0x69, 0x63, 0x73)
+	//if err != nil {
+	//return err
+	//}
 	err = en.WriteMapHeader(uint32(len(z.Metrics)))
 	if err != nil {
 		return
@@ -316,19 +316,19 @@ func (z *Span) EncodeMsg(en *msgp.Writer) (err error) {
 		}
 	}
 	// write "parent_id"
-	err = en.Append(0xa9, 0x70, 0x61, 0x72, 0x65, 0x6e, 0x74, 0x5f, 0x69, 0x64)
-	if err != nil {
-		return err
-	}
+	//err = en.Append(0xa9, 0x70, 0x61, 0x72, 0x65, 0x6e, 0x74, 0x5f, 0x69, 0x64)
+	//if err != nil {
+	//return err
+	//}
 	err = en.WriteUint64(z.ParentID)
 	if err != nil {
 		return
 	}
 	// write "type"
-	err = en.Append(0xa4, 0x74, 0x79, 0x70, 0x65)
-	if err != nil {
-		return err
-	}
+	//err = en.Append(0xa4, 0x74, 0x79, 0x70, 0x65)
+	//if err != nil {
+	//return err
+	//}
 	err = en.WriteString(z.Type)
 	if err != nil {
 		return
```

**Comment by knusbaum** [comment]
> I managed to get this over your PR:
> 
> ```
> $ benchcmp pr.bench trim.bench 
> benchmark                    old ns/op     new ns/op     delta
> BenchmarkDecodeRequest-8     66728632      60679977      -9.06%
> 
> benchmark                    old allocs     new allocs     delta
> BenchmarkDecodeRequest-8     501007         501007         +0.00%
> 
> benchmark                    old bytes     new bytes     delta
> BenchmarkDecodeRequest-8     102551545     93655019      -8.68%
> ```
> 
> By simplifying the codec a bit (this requires tracer changes): [cc5ecd1](https://github.com/DataDog/datadog-agent/commit/cc5ecd1d2fc99e5624da0f33f04a46be285d3ae8). This might be an easy win and change everywhere.

That looks great. Will we need to support the old format as well, though?
Maybe we can make this a new version of the traces endpoint. Then we can make whatever changes to the msgpack structure we want.

**Comment by gbbr** [comment]
Definitely a new endpoint, yes.

**Review by gbbr — COMMENTED** [review]
Because of the higher memory usage, this is only going to be successful if we make further improvements to the codec. This IMHO involves fine-tuning the wire format by [removing string keys](https://github.com/DataDog/datadog-agent/commit/cc5ecd1d2fc99e5624da0f33f04a46be285d3ae8) among with other potential ideas to reduce string duplication. For this reason I suggest that if we are to pursue this, do it in a new endpoint (`v0.5`) so that it actually becomes an option and a reality (breaking noone)

**Comment by knusbaum** [comment]
I am going to close this for now. There is a lot of work left to do to make direct decoding viable, and with improvements like https://github.com/DataDog/datadog-agent/pull/5965, it might prove to be unnecessary. We can revisit this in the future.

**Pull request closed without merging**